### PR TITLE
Fixed a data loss issue on lite member promotion

### PR DESCRIPTION
**Pull request by petrpleshachkov**
Notify cluster members on lite member promotion to update
memberGroupSize. Otherwise, a data member might be not aware of other
promoted data members, and it may cause backup operations
not being issued to other data members.

Closes https://github.com/hazelcast/hazelcast/issues/17621

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java [hunk 1]
```diff
@@ -367,6 +367,7 @@ private MemberImpl createNewMemberImplIfChanged(MemberInfo newMemberInfo, Member
                 member = clusterService.promoteAndGetLocalMember();
             } else {
                 member = createMember(newMemberInfo, member.getAttributes());
+                node.partitionService.memberAdded(member);
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java:370]: Shouldn't this be out of `else` block? What if this is on local member?

Comment (petrpleshachkov) [reply hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java:370]: On the local member the  `node.partitionService.memberAdded(member);` is doing update only when the node is not local. For local case, it issues a special task only when the node is a master. But the master is notified about promotion without this change. This is why I decided that it is enough to put it in the `else`.

Comment (petrpleshachkov) [reply hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java:370]: Also, the problem is when the member is not aware of **other**  members being promoted. That case is covered in the `else` branch

Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java:370]: I think issue is a non-master member is not aware of any member being promoted, including itself. So maybe we should modify `node.partitionService.memberAdded(member)` to update `memberGroupSize even with local member.

See this modified test:

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java [hunk 2]
```diff
@@ -368,6 +376,7 @@ private MemberImpl createNewMemberImplIfChanged(MemberInfo newMemberInfo, Member
             } else {
                 member = createMember(newMemberInfo, member.getAttributes());
             }
+            node.partitionService.memberAdded(member);
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java:379]: Not needed anymore?

**Review by mdogan — APPROVED** [review]

File: hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/PromoteLiteMemberTest.java [hunk 3]
```diff
@@ -323,6 +326,86 @@ public void promotion_shouldFail_whenMasterIsSuspected_duringPromotion() throws
...
+        secondHazelcastInstance.getLifecycleService().terminate();
+
+        // backup count for the map is set to 1
+        // even with 1 node down, no data loss is expected
+        assertTrueEventually(() -> assertEquals(entryCount, firstHazelcastInstance.getMap(mapName).size()));
+        assertTrueEventually(() -> assertEquals(entryCount, thirdHazelcastInstance.getMap(mapName).size()));
+    }
+
+    @Test
+    public void lite_member_promotion_data_loss2() throws InterruptedException {
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/PromoteLiteMemberTest.java:371]: same here.

File: hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/PromoteLiteMemberTest.java [hunk 4]
```diff
@@ -323,6 +326,86 @@ public void promotion_shouldFail_whenMasterIsSuspected_duringPromotion() throws
         }
     }
 
+    @Test
+    public void lite_member_promotion_data_loss() throws InterruptedException {
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/PromoteLiteMemberTest.java:330]: can we give a better name to test? Seeing the state tested against and expected behavior in test method name would be helpful for the future readers.

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java [hunk 5]
```diff
@@ -342,6 +346,10 @@ void updateMembers(MembersView membersView) {
 
         setMembers(MemberMap.createNew(membersView.getVersion(), members));
 
+        if (updatedLiteMember) {
+            updateMembersGroupSize();
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java:350]: it is hard to to follow similar named methods maybe we can make `updateMemberGroupsSize()` method public and can call it here: `node.partitionService.getPartitionStateManager().updateMemberGroupsSize()`

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java [hunk 6]
```diff
@@ -305,11 +305,15 @@ void updateMembers(MembersView membersView) {
...
         MemberImpl[] members = new MemberImpl[membersView.size()];
         int memberIndex = 0;
+        boolean updatedLiteMember = false;
         for (MemberInfo memberInfo : membersView.getMembers()) {
             Address address = memberInfo.getAddress();
             MemberImpl member = currentMemberMap.getMember(address);
 
             if (member != null && member.getUuid().equals(memberInfo.getUuid())) {
+                if (member.isLiteMember()) {
+                    updatedLiteMember = true;
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java:315]: can you please add a short comment to the code to explain the meaning of `updatedLiteMember`?

**Comment by mdogan** [comment]
@petrpleshachkov: This fix should be backported 4.0.z and 3.12.z branches.

**Comment by petrpleshachkov** [comment]
@ahmetmircik , I've addressed the comments.

**Review by ahmetmircik — APPROVED** [review]

**Comment by petrpleshachkov** [comment]
Guys, thanks for the review, I'm merging the PR.

**Comment by mmedenjak** [comment]
@petrpleshachkov have you backported the PR?

**Pull request merged**